pkg/mcs/resource_manager/server/token_bukets.go

```diff
@@ -38,107 +41,259 @@ type GroupTokenBucket struct {
+	gtb.Tokens = state.Tokens
+	gtb.LastUpdate = state.LastUpdate
+	gtb.Initialized = state.Initialized
+}
+
+// TokenSlot is used to split a token bucket into multiple slots to
+// server different clients within the same resource group.
+type TokenSlot struct {
+	// settings is the token limit settings for the slot.
+	settings *rmpb.TokenLimitSettings
+	// requireTokensSum is the number of tokens required.
+	requireTokensSum float64
+	// tokenCapacity is the number of tokens in the slot.
+	tokenCapacity     float64
+	lastTokenCapacity float64
+
+	lastUpdate *time.Time
+}
+
 // GroupTokenBucketState is the running state of TokenBucket.
 type GroupTokenBucketState struct {
-	Tokens      float64    `json:"tokens,omitempty"`
+	Tokens float64 `json:"tokens,omitempty"`
+	// ClientUniqueID -> TokenSlot
+	tokenSlots                 map[uint64]*TokenSlot
+	clientConsumptionTokensSum float64
+
 	LastUpdate  *time.Time `json:"last_update,omitempty"`
 	Initialized bool       `json:"initialized"`
 	// settingChanged is used to avoid that the number of tokens returned is jitter because of changing fill rate.
 	settingChanged bool
 }
 
 // Clone returns the copy of GroupTokenBucketState
-func (s *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+func (gts *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+	tokenSlots := make(map[uint64]*TokenSlot)
+	for id, tokens := range gts.tokenSlots {
+		tokenSlots[id] = tokens
+	}
+	var lastUpdate *time.Time
+	if gts.LastUpdate != nil {
+		newLastUpdate := *gts.LastUpdate
+		lastUpdate = &newLastUpdate
+	}
 	return &GroupTokenBucketState{
-		Tokens:      s.Tokens,
-		LastUpdate:  s.LastUpdate,
-		Initialized: s.Initialized,
+		Tokens:                     gts.Tokens,
+		LastUpdate:                 lastUpdate,
+		Initialized:                gts.Initialized,
+		tokenSlots:                 tokenSlots,
+		clientConsumptionTokensSum: gts.clientConsumptionTokensSum,
+	}
+}
+
+func (gts *GroupTokenBucketState) cleanupAssignTokenSum(cleanToken bool) {
+	gts.clientConsumptionTokensSum = 0
+	for _, slot := range gts.tokenSlots {
+		slot.requireTokensSum = 0
+		if cleanToken {
+			slot.tokenCapacity = 0
+			slot.lastTokenCapacity = 0
+		}
 	}
 }
 
+func (gts *GroupTokenBucketState) balanceSlotTokens(
+	clientUniqueID uint64,
+	settings *rmpb.TokenLimitSettings,
+	requiredToken, elapseTokens float64) {
+	now := time.Now()
+
+	slot, exist := gts.tokenSlots[clientUniqueID]
+	if !exist {
+		if requiredToken == 0 {
+			return
+		}
+		gts.cleanupAssignTokenSum(false)
+		slot = &TokenSlot{}
+		gts.tokenSlots[clientUniqueID] = slot
+	} else if gts.clientConsumptionTokensSum >= maxAssignTokens {
+		gts.cleanupAssignTokenSum(false)
+	}
+
+	// Clean up slot that have not been used for defaultSlotStalePeriod.
+	if requiredToken == 0 && now.Sub(*slot.lastUpdate) >= defaultSlotStalePeriod {
+		delete(gts.tokenSlots, clientUniqueID)
+		gts.cleanupAssignTokenSum(false)
```

Comment:
IMO, the tokens of the deleted resource group should be balanced, whether sufficient or liability.

Reply:
If the client is down, it will always affect other ru allocations.
For example, if 1 of the 3 machines goes down, the total qps will be reduced by 1/3

Reply:
We should consider behavior in severless cluster. Tokens can be very numerous